Blackjack/Game/BlackjackGame.m

```diff
@@ -0,0 +1,262 @@
+      Player *player = [[Player alloc] initWithName:name chips: MINIMUM_BUYIN delegate:self];
+      [newPlayers addObject: player];
+    }
+    _players = newPlayers;
+    _currentPlayer = newPlayers.firstObject;
+    _deck = deck;
+    _delegate = delegate;
+    _bets = [NSMutableDictionary new];
+    _dealer = [[Player alloc] initWithName:@"Dealer" chips: 0 delegate:self];
+    _state = IN_GAME;
+  }
+  return self;
+}
+
+- (void)setState:(GameState)state {
+  _state = state;
+  [self.delegate handleChangesforNewState: state];
+}
+
+// MARK: - Player Filters
+
+- (NSInteger)getBetAmountForPlayer:(Player *)player {
+  return [self.bets valueForKey:player.identifier].integerValue;
+}
+
+- (NSArray<Player *> *)currentActivePlayers {
+  NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(Player *player, NSDictionary *bindings) {
+    return player.state == PLAYING;
+  }];
+  return [self.players filteredArrayUsingPredicate:predicate];
+}
+
+- (NSArray<Player *> *)playersAvailableForNextFound {
+  NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(Player *player, NSDictionary *bindings) {
+    return player.chips >= MINIMUM_ROUND_BET;
+  }];
+  return [self.players filteredArrayUsingPredicate:predicate];
+}
+
+// MARK: - Game Logic
+
+- (void)collectDecision:(Decision)decision {
+  BOOL isLastPlayer = self.currentPlayer == [self currentActivePlayers].lastObject;
+  
+  switch (decision) {
+    case HIT: {
+      [self.currentPlayer acceptNewCard:[self.deck drawRandomCardWithFaceUp:YES]];
+      
+      if (self.currentPlayer.state == PLAYING) {
+        return;
+      }
+      break;
+    }
+      
+    case SURRENDER:
+      [self.currentPlayer setState:BUST];
+      
+    case STAND: {
+      break;
+    }
+  }
+  
+  if (isLastPlayer) {
+    [self finlizeRound];
+  } else {
+    [self nextPlayer];
+  }
+}
+
+- (void)collectBet:(NSInteger)amount {
+  if (self.currentPlayer.chips >= amount) {
+    [self.bets setObject:[NSNumber numberWithInteger:amount] forKey:self.currentPlayer.identifier];
+    [self.currentPlayer collectBet:amount];
+    [self nextPlayer];
+  } else {
+    NSAssert(self.currentPlayer.chips < amount, @"Player should never get to this point");
+  }
+  
+  if (self.bets[self.players.lastObject.identifier]) {
+    [self dealCardsFaceUp:YES];
+    [self dealCardsFaceUp:NO];
+    
+    if (self.currentPlayer == self.players.firstObject && self.currentPlayer.state != PLAYING) {
+      [self nextPlayer];
+    }
+    self.state = BETS_OVER;
+  }
+}
+
+-(void)dealCardsFaceUp:(BOOL)isFaceUp {
+  for (Player *player in self.players) {
+    BOOL _isFaceUp = player == self.players.firstObject ? YES : isFaceUp;
+    [player acceptNewCard:[self.deck drawRandomCardWithFaceUp:_isFaceUp]];
+  }
+  [self.dealer acceptNewCard:[self.deck drawRandomCardWithFaceUp:isFaceUp]];
+}
+
+-(void)nextPlayer {
+  NSMutableArray<NSNumber *> *playersIndexToUpdated = [NSMutableArray new];
```

Comment:
This variable defines 2 problems been resolved by the same variable..

On the one hand you have the need to focusOnAPlayer if needed (You don't need the `playersIndexToUpdated` to do that).

Oh the other, the need to update the UI for each player.

I think adding those 2 players can be done outside the `if else` flow and it will create a much cleaner code (hint: think about the mathematical operation of modulo number of players)

Reply:
In addition the `focusOn` is always called on the new currentPlayer no? :)

Reply:
Logic changed